### PR TITLE
[Reviewer: AJH] Add HTTP Lazy Targeting UTs

### PR DESCRIPTION
**Pull request by tomwatson1024**
Add unit tests, and modify existing unit tests, for HTTP lazy targeting in cpp-common pull request [#549](https://github.com/Metaswitch/cpp-common/pull/549).

### CODE REVIEW

File: src/ut/baseresolver_test.cpp [hunk 1]
```diff
@@ -95,6 +98,14 @@ class BaseResolverTest : public ResolverTest
     return targets;
   }
 
+  /// Implements the resolve_iter mathod using a BaseResolver
```
Comment (AlexHockey) [inline src/ut/baseresolver_test.cpp]: mathod

Comment (AlexHockey) [inline src/ut/baseresolver_test.cpp]: This is just a helper method, right? The comment makes it sound like you are implementing an interface but I don't think that's the case.

File: src/ut/baseresolver_test.cpp [hunk 2]
```diff
@@ -140,7 +151,7 @@ TEST_F(BaseResolverTest, IPv4AddressResolutionManyTargets)
 TEST_F(BaseResolverTest, ARecordAtLeastOneGray)
 {
   add_white_records(DEFAULT_COUNT);
-  AddrInfo gray_record = ip_to_addr_info("3.0.0.0");
+  AddrInfo gray_record = ResolverTest::ip_to_addr_info("3.0.0.0");
```
Comment (AlexHockey) [inline src/ut/baseresolver_test.cpp:154]: Hmmm, why do you need to qualify this now?

File: src/ut/baseresolver_test.cpp [hunk 3]
```diff
@@ -305,6 +316,51 @@ TEST_F(BaseResolverTest, ARecordMakeUpUsedGray)
...
+  add_white_records(1);
+  BaseAddrIterator* it = resolve_iter();
+
+  // The value of record should be set by the iterator.
+  EXPECT_TRUE(it->next(record));
+  EXPECT_EQ(record, expected_record);
+
+  // The value of record should be left unchanged by the iterator.
+  EXPECT_FALSE(it->next(record));
+  EXPECT_EQ(record, expected_record);
```
Comment (AlexHockey) [inline src/ut/baseresolver_test.cpp:335]: This second set of tests should be a separate testcase.

Comment (AlexHockey) [inline src/ut/baseresolver_test.cpp:335]: Or maybe keep this the way it is and have a separate testcase that just tests when there are no targets to return.

File: src/ut/baseresolver_test.cpp [hunk 4]
```diff
@@ -305,6 +316,51 @@ TEST_F(BaseResolverTest, ARecordMakeUpUsedGray)
...
+  // Move the record to the graylist
+  cwtest_advance_time_ms(31000);
+
+  // The record should be returned by the next call to the iterator
+  AddrInfo record;
+  EXPECT_TRUE(it->next(record));
+  EXPECT_EQ(record, black_to_gray_record);
+
+  delete it; it = nullptr;
+}
```
Comment (AlexHockey) [inline src/ut/baseresolver_test.cpp:459]: You could make this test more convincing by creating two iterators and consuming one before you advance time and the other afterwards. I know the first iterator could return an acceptable address by chance, but this testcase would make it really clear what the point of lazy iteration is.

**Comment by AlexHockey** [comment]
Thanks Tom. The tests you've added look pretty good. However, I think there are a few testcases it would be good to add, for the BaseAddrIterator at least. Hopefully these won't take too long to add?
- Call next() on an iterator with no results. 
- Call take() when there are no results. 
- take() all the results. Calling take() a second time returns no results. 
- take() some of the results. Calling take() a second time returns more. 
- take() more results than are available. 
- Mix next() and take() calls.

File: src/ut/baseresolver_test.cpp (hunk 2 above)
Comment (tomwatson1024) [reply src/ut/baseresolver_test.cpp:154]: I have changed the method to be static, as there's no need for a ResolverTest object.

Comment (AlexHockey) [reply src/ut/baseresolver_test.cpp:154]: Ah, OK.

**Comment by AlexHockey** [comment]
:+1:

**Pull request merged**